feat(accounts): add APIs to manage subscriptions and report capabilities

lmorchard opened the pull request:
TODOs:
- [x] Try to resolve TODOs around subhub integration - made some assertive assumptions, but issue #719 exists to finish this for real
- [x] Add OAuth token strategy to subscription CRUD methods for use by payment & management page on separate domain (#715)
- [x] Auth server method to issue OAuth token for payment pages from session token
- [x] Ensure all subscriptions are cancelled when account is deleted
- [x] Metrics events? (Present, but most probably incorrect in absence of a metrics spec.)
- [x] Customs server calls? (Present, but not sure if correct.)

~This is an embarassingly early draft pull request. But I'm opening it to let folks see where I'm at with auth (#714) & profile API (#708) additions that will be used by payment pages (epic #695) and RPs (epic #706).~

I'm of course also new to all these bits of the stack, so any & all advice & critique are gratefully welcomed 😅

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,188 @@
+      },
+      handler: async function (request) {
+        log.begin('getSubscriptionCapabilities', request);
+        // TODO: is this where to get client_id?
```
I want to use the ID of the client for which an OAuth token was issued to determine what subscription capabilities are visible.

How do I get that ID? Is it already in `request.auth.credentials`? Or will I have to bounce the token off an OAuth server API? I'm going to dig around to learn this myself, but someone might know this faster.

philbooth replied:
If you have an `oauthToken` then yep, I think (?) it's a property on that. It won't be set if you have a `sessionToken` though.

philbooth replied:
(btw, in case it wasn't clear, these strategies just end up returning instances of the token objects from `lib/tokens`, you can see how they're made with `makeCredentialFn` in `lib/server.js`)

philbooth commented on packages/fxa-auth-server/lib/db.js:
```diff
@@ -1358,6 +1358,58 @@ module.exports = (
   };
 
+  SAFE_URLS.createAccountSubscription = new SafeUrl(
+    '/account/:id/subscriptions/:subscriptionId',
```
Typo: this `:id` in the safe URL should be `:uid` to match the argument to `pool.put` (or vice versa).

lmorchard replied:
Oof yeah, I made that typo all over the place :/

vbudhram commented on packages/fxa-auth-server/test/local/routes/subscriptions.js:
```diff
@@ -0,0 +1,409 @@
+      }
+    });
+
+    it('should correctly handle payment token rejection', async () => {
```
Can users create multiple subscriptions to the same service? Or would that fail?

lmorchard replied:
Hmm, that's a good question - I think that's something we might need to enforce at the subhub level

vbudhram commented on packages/fxa-auth-server/test/local/routes/subscriptions.js:
```diff
@@ -0,0 +1,409 @@
+      }
+    });
+
+    it('should correctly handle payment token rejection', async () => {
```
You can also testing for updating an unknown subscription.

lmorchard replied:
This method's just for creating a subscription, so we're really just dealing with a plan & payment here so far.

lmorchard commented on packages/fxa-auth-server/lib/error.js:
```diff
@@ -1033,6 +1036,39 @@ AppError.invalidPkceChallenge = (pkceHashValue) => {
+
+AppError.unknownSubscriptionPlan = (planId) => {
+  return new AppError({
+    code: 400,
```
No, subscriptions & plans are different entities. A plan is what a user subscribes to in order to access a product.

philbooth commented on packages/fxa-auth-server/config/index.js:
```diff
@@ -569,6 +569,54 @@ const conf = convict({
+    enabled: {
+      doc: 'Indicates whether talking to the SubHub server is enabled',
+      format: Boolean,
+      default: false,
```
Should the default value for this be `true`, with the `false` override in `dev.json`? I think Ops have a preference that the production config uses as many defaults as possible.

lmorchard replied:
I made the default false since we are likely to merge & deploy this stuff in advance of any subhub installations being available. Figured that would be safer, but maybe we can ensure it's false along with the deployment

philbooth commented on packages/fxa-auth-server/config/index.js:
```diff
@@ -569,6 +569,54 @@ const conf = convict({
+    key: {
+      doc: 'Authentication key to use when accessing SubHub server',
+      format: String,
+      default: 'Correct_Horse_Battery_Staple_1',
```
Not sure if it's well-established enough to be considered a convention, but elsewhere we have used the literal string `YOU MUST CHANGE ME` for these. Also not sure if that actually does anything to reduce the chance of it not being set, but mentioned fwiw. If nothing else it's a clear signal for self-hosters.

philbooth replied:
(although self-hosters won't need to mess with this one of course, so maybe it's a moot point)

lmorchard replied:
I just copied this directly from the defaults for pushbox, since I based the subhub module on it. But I can change it to `YOU MUST CHANGE ME` - doesn't much matter here since there's likely not going to be a public repo with a self-hostable subhub

philbooth commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -3008,6 +3020,42 @@ Returns SMS status for the current user.
+#### GET /subscriptions/plans
+
+:lock: HAWK-authenticated with session token
+<!--begin-route-get-subscriptionplans-->
```
Just FYI, these HTML comments are a hangover from when this doc used to be auto-generated. But we ditched that recently because the script kept braking, so we can stop adding them now.

philbooth commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,193 @@
+
+        // TODO: need a metrics spec
+        request.emitMetricsEvent(
+          'subscriptions.update_payment',
```
Not a massive deal, but the events we send to `emitMetricsEvent` mostly use camel case. We try to keep them consistent to make writing queries easier in Redash. There are a couple of events that use snake case, but those are exceptions that snuck through rather than the general rule.

The amplitude events, which are specified elsewhere (see `lib/metrics/amplitude.js`), *do* use snake case because that schema was defined fresh in isolation from the rest of our event architecture.

philbooth commented on packages/fxa-auth-server/test/local/routes/account.js:
```diff
@@ -1598,10 +1604,24 @@ describe('/account/destroy', () => {
+    mockDB.fetchAccountSubscriptions =
+      sinon.spy(async (uid) => expectedSubscriptions);
+    mockDB.deleteAccountSubscription =
+      sinon.spy(async (uid, subscriptionId) => true);
```
You're welcome to add these to the method list for `mockDB` if you like, to save defining them inline in the test body like this. I realise they're not the most broadly useful methods for mocking, but neither is a lot of the other stuff in there and the less setup code in the tests the better, I reckon.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,193 @@
+        },
+        validate: {
+          payload: {
+            token: isA.string().required()
```
FWIW, this is defined by the Stripe API, and I can't find any promise of max length in their API docs so far.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,193 @@
+    {
+      method: 'GET',
+      path: '/subscriptions/plans',
+      options: {
```
I added some tenative validation anyway, we can update it when TBD is D

lmorchard commented:
Okay, I think I'm biting the bullet and taking the "draft" status off this PR. Anything else that needs doing I think either needs to surface from review and / or punt off into follow-up PRs. Also going to highlight some bits where I think I need help and where I made guesses

lmorchard commented on packages/fxa-auth-server/config/dev.json:
```diff
@@ -35,4 +35,47 @@
+    }
+  },
+  "subscriptions": {
+    "managementClientId": "3c49430b43dfba77",
```
Do we need to create a new OAuth client specifically for issuing the tokens for the off-domain subscriptions management pages?

lmorchard replied:
This is where I kind of cheated to get tests passing - this ID is the "Android Components Reference Browser" client used in `test/remote/oauth_tests.js`. Probably needs to be a completely new client

lmorchard replied:
I think I'm going to file a follow-up issue to address this in the content server work

lmorchard commented on packages/fxa-auth-server/config/index.js:
```diff
@@ -569,6 +569,66 @@ const conf = convict({
+  },
+  subscriptions: {
+    managementClientId: {
+      doc: 'OAuth client ID for management pages',
```
Making the OAuth client ID for off-domain subscription management pages configurable. Kind of guessing it's necessary.

lmorchard replied:
So, I guess the thing I don't know here is: What's the process for creating a new OAuth client, and after doing so can I use that ID here for `dev.json` and remote tests? (Would it basically be the ["Becoming a Firefox Accounts relier"](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction#Becoming_a_Firefox_Accounts_relier) process from MDN?)

lmorchard replied:
Though, actually, if I'm getting rid of this route... the place where we'll need to deal with that client ID is in the client - i.e. the payment pages on the content server. That is, unless restricting access to the subscription management scope requires the client ID

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,268 @@
+  return [
+    {
+      method: 'POST',
+      path: '/subscriptions/token',
```
This is how & where I've made an attempt at issuing the OAuth token for use by the off-domain subscription management pages. I copied it in part from the `/oauth/token` but with hardcoded values. Could use some scruitiny here to see if this could be done better.

FWIW, I'm imagining it'll be called from 1) FxA subscription settings page when the "manage" button is clicked for subscriptions and 2) on an FxA page linked-to from external products that will redirect to a subscription signup page.

lmorchard replied:
Yeah, that would simplify things here. I thought maybe having a special-purpose route just for issuing this particular token would be a good idea, but it simplifies things not to have it.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,268 @@
+  const totpUtils = require('./utils/totp')(log, config, db);
+
+  const SUBSCRIPTIONS_MANAGEMENT_SCOPE =
+    'https://identity.mozilla.com/apps/subscriptions';
```
I *think* I read somewhere that we want scopes to be in URL form going forward. So, I made this one up for the subscription management pages. Is this the right thing to do?

lmorchard replied:
Well, this scope is specifically for access to the subscriptions CRUD methods. But, on the payment pages we also want to read profile details for display. So, I figured we'd need both here.

lmorchard commented on packages/fxa-auth-server/test/remote/subscriptions_tests.js:
```diff
@@ -0,0 +1,50 @@
+  });
+
+  it('successfully grants tokens for subscriptions management', async () => {
+    const res = await client.api.grantSubscriptionsManagementToken(client.sessionToken);
```
Really minimal test here. Doesn't assert anything about IP binding or the like. Suggestions more than welcome!

lmorchard replied:
Just going to remove this route entirely.

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -1195,72 +1227,78 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
         const form = request.payload;
         const authPW = form.authPW;
 
-        return customs.check(request, form.email, 'accountDestroy')
```
Feeling a bit dangerous here: While adding code to ensure all subscriptions get cancelled along with account deletion, I took a shot at converting the whole method to async / await. I think I covered everything, but can convert it back to plain promises if it seems hinky at all

lmorchard replied:
Actually, nevermind: Looks like I did break something in reworking this chunk of code, so I fell back to just adding a new bit in the promise chain to just do what I needed to cancel subscriptions.

lmorchard commented:
I may also be tracking down a few confusing failures in remote auth-server tests I'm seeing locally. 😓

shane-tomlinson commented:
This just might be one of the largest first major contributions I have ever seen! 👀 

<img width="159" alt="Screenshot 2019-04-17 at 10 50 50" src="https://user-images.githubusercontent.com/848085/56278779-c257bb80-60fe-11e9-97c3-cd4d847b646b.png">

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+        const paymentResult = await subhub.createSubscription(uid, token, planId);
+
+        const subscriptionId = paymentResult.sub_id;
+        await db.createAccountSubscription({
```
The create, update, and delete operations all update an account's profile information and should notify attached devices/services of the profile update. See https://github.com/mozilla/fxa/blob/0d52b0fd6ea8945203e47e4784754e3e8b8b65fc/packages/fxa-auth-server/lib/routes/emails.js#L738:L742

In this case instead of emitting a [`primaryEmailChanged` event](https://github.com/mozilla/fxa/blob/0d52b0fd6ea8945203e47e4784754e3e8b8b65fc/packages/fxa-auth-server/lib/routes/emails.js#L739) to attached services, we should emit a [`profileDataChanged` event](https://github.com/mozilla/fxa/blob/0d52b0fd6ea8945203e47e4784754e3e8b8b65fc/packages/fxa-auth-server/lib/routes/totp.js#L297). Of course, there may be some new subscription events that should be emit instead, maybe @bbangert has some input here.

lmorchard replied:
I'll add this general `profileDataChanged` event to to the subscriptions routes.

> there may be some new subscription events that should be emit instead

I think this is basically issue #701 though

shane-tomlinson commented on packages/fxa-auth-server/lib/subhub.js:
```diff
@@ -0,0 +1,315 @@
+          log.error('subhub.listSubscriptions.1', { uid, err });
+          // TODO: update with subhub listSubscriptions error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
```
`unknownAccount` seems a bit strange here because the account is known to *us* otherwise this method wouldn't be called, it's just not known by subhub. Maybe another [new error number is warranted](https://github.com/mozilla/fxa/pull/819/files#diff-fd7dff0f3afc526facd6365fed66bcb7R1039) to indicate "no known subscriptions".

lmorchard replied:
Yeah, [this is an open question](https://github.com/mozilla/subhub/issues/3) I have on subhub: Is there a difference between "no known subscriptions" or "never been a customer yet" or "was a customer, cancelled all subscriptions" - but are all of them the same error / response?

lmorchard replied:
For now, I created a new `unknownCustomer` error to capture this. It may change, depending on what subhub eventually does (issue #719 )

shane-tomlinson commented on packages/fxa-auth-server/lib/subhub.js:
```diff
@@ -0,0 +1,315 @@
+          log.error('subhub.cancelSubscription.1', { uid, sub_id, err });
+          // TODO: update with subhub cancelSubscription error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
```
Similar to the comment above, it seems like the uid is known to *us* otherwise this would not be called, just not to subhub. Maybe this warrants a new error number to indicate "no known subscriptions".

lmorchard replied:
Sure, I can create a new error for this in particular. Can always tweak it later as subhub materializes

shane-tomlinson commented on packages/fxa-auth-server/lib/subhub.js:
```diff
@@ -0,0 +1,315 @@
+      try {
+        return await api.listSubscriptions(uid);
+      } catch (err) {
+        if (err.statusCode === 404) {
```
Is this status code guaranteed to be fixed? Is a status code/message combo how subhub defines their errors?

lmorchard replied:
The answers to both of these are 🤷‍♂️ TBD. This part of the subhub spec is not yet finished.

lmorchard replied:
Yeah, I think this is a part of issue #719

shane-tomlinson commented on packages/fxa-auth-server/lib/subhub.js:
```diff
@@ -0,0 +1,315 @@
+    },
+
+    async cancelSubscription(uid, sub_id) {
+      const key = subscriptionsKey(uid, sub_id);
```
Some of the errors possible in the real implementation are not possible in the stub, will that be a hindrance to testing?

lmorchard replied:
Possibly, I can give this another pass to try to make it closer to my guesses at subhub

lmorchard replied:
On further thought, I think this might be worth expanding on as part of #719 when we're more certain about the actual error that will come out of subhub. Going to add a note to that issue.

lmorchard replied:
https://github.com/mozilla/fxa/issues/719#issuecomment-484695091

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+  const totpUtils = require('./utils/totp')(log, config, db);
+
+  const SUBSCRIPTIONS_MANAGEMENT_SCOPE =
+    'https://identity.mozilla.com/apps/subscriptions';
```
I imagine we'll want to limit the ability of this scope to be requested by arbitrary RPs. Perhaps we only allow the content server and/or the subscription server to request this scope and we can relax the rule if necessary later.

lmorchard replied:
Yeah, this scope should only ever really be granted to the payment pages. I'm not sure how to restrict that, though.

lmorchard replied:
FWIW, I think this is worth punting off to a follow up issue:

https://github.com/mozilla/fxa/issues/881

shane-tomlinson commented on packages/fxa-profile-server/lib/routes/_core_profile.js:
```diff
@@ -85,6 +86,9 @@ module.exports = {
       if (typeof body.authenticatorAssuranceLevel !== 'undefined') {
         result.twoFactorAuthentication = body.authenticatorAssuranceLevel >= 2;
       }
+      if (typeof body.subscriptions !== 'undefined') {
```
Can we only return this if an appropriate scope is set? It makes me nervous that any RP can get a list of all the subscriptions that are set for a relier.

If we guard the subscription list, we will need a way to surface whether the user has *any* active subscriptions to other RPs, e.g., SUMO doesn't care about the actual subscription list for a user, they just want to know whether the user has at least one active subscription.

lmorchard replied:
> It makes me nervous that any RP can get a list of all the subscriptions that are set for a relier.

Well, each RP should only be shown the subset of subscription capabilities listed for them in `clientCapabilities` in the server config. That's meant to guard visibility of subscription status.

> we will need a way to surface whether the user has any active subscriptions to other RPs

It occurred to me today that I might add a few "special" products like `default` that every registered user gets regardless of subscriptions and `defaultSubscribed` that users with at least one subscription to any product gets. Then, we should be able to list those in `clientCapabilities` for a client like SUMO.

lmorchard replied:
I added the special products for any registered user and any user with a subscription. Also, I think the restricted capability filtering by client ID takes care of guarding subscriptions visibility here.

shane-tomlinson commented on packages/fxa-profile-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,36 @@
+module.exports = {
+  auth: {
+    strategy: 'oauth',
+    // TODO: using email scope for now, may change someday
```
I hope `someday` is `today` or at least `before this PR lands`.

lmorchard replied:
I think this was something @bbangert mentioned - i.e. to re-use the profile email scope for access to read subscription status. We can change it though, if there's a different existing scope or we want to invent a new one.

lmorchard replied:
FWIW: If only in the interests of getting this landed to unblock other work, I'm thinking it might be okay to punt this to a follow up issue to be resolved before the launch of the first RP for this new subscriptions stuff.

lmorchard replied:
Created a follow-up issue for this - easy enough to change when we want to change it, as long as we do it before the first subscription RP joins on

https://github.com/mozilla/fxa/issues/882

lmorchard commented on packages/fxa-auth-server/config/index.js:
```diff
@@ -569,6 +569,66 @@ const conf = convict({
+      default: {}
+    },
+    clientCapabilities: {
+      doc: 'Mappings from OAuth client IDs to relevant subscription capabilities',
```
Not interchangeable. `productCapabilities` lists out what subscription capabilities a user gains when they have a subscription to a product. `clientCapabilities` lists what capabilities are relevant to each given client.

So, when a client fetches a profile, they see a subset of the user's fill capabilities (`productCapabilities`) that match up with the client-relevant capabilities (`clientCapabilities`). It's a kind of hardcoded (well, server-configured) access control that only lets RPs see the capabilities related to their offerings.

lmorchard replied:
Hopefully this is a little more clear now that I've split out this particular algorithm into `lib/routes/utils/subscriptions.js' and added a unit test to spell out a few cases.

lmorchard commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -76,6 +76,12 @@ see [`mozilla/fxa-js-client`](https://github.com/mozilla/fxa-js-client).
     * [POST /sms (:lock: sessionToken)](#post-sms)
     * [GET /sms/status (:lock: sessionToken)](#get-smsstatus)
+  * [Subscriptions](#subscriptions)
+    * [GET /subscriptions/plans (:lock: sessionToken)](#get-subscriptionsplans)
```
Yeah, good catch - these definitely need an update since I started adding the OAuth token strategies

lmorchard replied:
Added a todo to ensure the docs are copacetic to this follow-up

https://github.com/mozilla/fxa/issues/887

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -816,48 +816,80 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
-                });
+        const account = await db.account(uid);
+
+        if (scope.contains('profile:email')) {
```
Didn't look like anything else in this route was using ScopeSet, so I tried to follow the existing scope checks. Is there something different here? Or should the other scope checks use ScopeSet too?

vladikoff replied:
ah 👍  , the `scope = { contains: () => true };` threw me off

lmorchard replied:
Yeah, it confused me for a second two, I couldn't remember exactly what I'd done there :)

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -816,48 +816,80 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
+          if (subscriptions.length > 0) {
+            const capabilitiesToReveal = new Set();
+            const clientVisibleCapabilities = clientCapabilities[client_id] || [];
+            for (const subscription of subscriptions) {
```
I had this written using map / filter at one point, but figured a plain for loop looked cleaner. But, I can switch back to a functional approach

lmorchard replied:
I switched to more array functional code in the new helper module.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+            plan_id: isA.string().required(),
+            product_id: isA.string().required(),
+            interval: isA.string().required(),
+            amount: isA.string().required(),
```
Looks like both stripe & subhub define this as a number, so I can change that

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+        await customs.check(request, email, 'createSubscription');
+
+        const planId = request.payload.plan_id;
+        const token = request.payload.token;
```
Yeah, I might change this to `pmt_token` since that's what subhub calls it

lmorchard replied:
I changed this to `paymentToken`

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+        const paymentResult = await subhub.createSubscription(uid, token, planId);
+
+        const subscriptionId = paymentResult.sub_id;
+        await db.createAccountSubscription({
```
That's a good question. If our `db.createAccountSubscription` call fails after `subhub.createSubscription` succeeds, we definitely won't want to delete the subhub subscription - the customer will have been charged money at that point.

Not sure what the best solution here is. We could attempt some retries? Alternatively, we could try to lean on subhub notifications (issue #713) as a fallback measure - those should be keeping us up to date when subscriptions expire, maybe they can also ensure we've stored record when a subscription starts.

lmorchard replied:
Tossing this to a follow-up because it could use some discussion

https://github.com/mozilla/fxa/issues/886

irrationalagent commented:
>     * Work with @irrationalagent on defining the event taxonomy for Amplitude.

In my queue - met with John to get his requirements. @lmorchard what's your timeline here? im a bit crunched this week, ok if I get you a first draft of a metrics spec early next week or are you looking to merge this ASAP?

lmorchard commented:
> looking to merge this ASAP?

@irrationalagent Yeah, want to get this merged ASAP to ensure other work doesn't get blocked. But, I'd say maybe we should just consider this a bit of a work in progress we can circle back to and iterate on in further PRs. I think it needs to land soon, but won't actually be in use for a decent amount of time yet.

irrationalagent commented:
> > looking to merge this ASAP?
> 
> @irrationalagent Yeah, want to get this merged ASAP to ensure other work doesn't get blocked. But, I'd say maybe we should just consider this a bit of a work in progress we can circle back to and iterate on in further PRs. I think it needs to land soon, but won't actually be in use for a decent amount of time yet.

cool, maybe we can work on a follow up metrics pr then.

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -816,48 +816,80 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
+            } = {}
+          } = config;
+
+          const subscriptions = await db.fetchAccountSubscriptions(uid) || [];
```
I just extracted this stuff out into a separate module and added some unit tests for it

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+        },
+        response: {
+          schema: isA.object().keys({
+            plan_id: isA.string().required(),
```
Sure, I can move them to the validators module. They're defined by subhub but practically speaking are carried over from the Stripe API, which only defines them as "string" and with vague promises of being less than 255 chars from a mailing list thread I found.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+        log.info('subscriptions.createSubscription.success', { uid, subscriptionId });
+
+        // TODO: need a metrics spec
+        request.emitMetricsEvent(
```
I'm stripping out the metrics code for now - we'll follow up on this with another PR rather than block and further complicate this one.

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -1219,6 +1251,19 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
+            const subscriptions = await db.fetchAccountSubscriptions(uid) || [];
+            for (const subscription of subscriptions) {
+              const { subscriptionId } = subscription;
+              await subhub.cancelSubscription(uid, subscriptionId);
```
We might be able to do a retry for the delete here. We can't re-subscribe the user with subhub, since that requires a freshly-authorized payment token from the Stripe widget (and thus a payment). 

This might be another place where we want to lean on events processing from subhub? I'm assuming that there should be a subscription-cancelled event that comes through, both for when it's cancelled via direct user action here and when cancelled via automatic expiration.

lmorchard replied:
Another question I guess is - do we have anywhere else that we do retries or otherwise try to recover from failures in the middle of some non-atomic DB API calls? Looking for an example I could steal

lmorchard replied:
Actually, going to resolve this one, since I tweaked the subscription cancelling to happen after password check. The failure case stuff is mentioned in another comment.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+      options: {
+        auth: {
+          strategies: [
+            'sessionToken',
```
Still trying to work through this one. My Hapi-knowledge is not strong, and I'm getting errors when I remove the sessionToken strategy: `Payload authentication requires at least one strategy with payload support in /v1/oauth/subscriptions/plans`. I'm also having trouble finding anywhere else in the auth server that just uses the `oauthToken` strategy as an example.

lmorchard replied:
I keep banging my head on removing the `sessionToken` strategy... Going to punt to a follow-up and continue working on it to hopefully get this thing landed.

https://github.com/mozilla/fxa/issues/887

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+
+        const { uid, email } = await handleAuth(request.auth);
+
+        await customs.check(request, email, 'createSubscription');
```
I tweaked this to omit `paymentToken` - the renamed version of `token`

lmorchard commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -315,6 +321,12 @@ for `code` and `errno` are:
   Public clients require PKCE OAuth parameters
 * `code: 400, errno: 171`:
   Required Authentication Context Reference values could not be satisfied
+* `code: 404, errno: 176`:
```
Added these errors, but I think I'll defer to a follow-up issue for the other missing errors.

lmorchard replied:
Filed a follow-up issue for this - https://github.com/mozilla/fxa/issues/885

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,272 @@
+      },
+      handler: async function (request) {
+        log.begin('subscriptions.listPlans', request);
+        await handleAuth(request.auth);
```
Maybe? Should be fairly static results for display on the management page though. If we do rate limit, is that a thing we do elsewhere or maybe in the customs server? (looking for an example)

vbudhram replied:
FWIW, I think it is ok to not rate limiting here since we are using an authentication strategy. If the route was unauthenticated, we should definitely rate limit though.

lmorchard replied:
Filed a follow-up for this - https://github.com/mozilla/fxa/issues/884

lmorchard commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -816,48 +816,80 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
+          res.authenticatorAssuranceLevel = authMethods.maximumAssuranceLevel(amrValues);
+        }
+
+        // TODO: email for now, but someday we may change scope for subscriptions
```
Going to resolve this in favor of the later comment about this scope

lmorchard commented:
> At least initially, I think we should only tell RPs which subscriptions a user has IFF the access token contains the subscription management scope. We can tell RPs whether the user has any active subscription using a simple boolean, but listing the actual subscriptions seems a step too far, e.g., I wouldn't want 3rd party service A knowing I have a subscription with 3rd party service B.

FWIW, just to circle back to this - we've got an algorithm in here to show clients only the subscription capabilities relevant to their offerings. So, I think we're covered on service A not seeing capabilities for service B - unless we've configured the list such that they share some of them.

lmorchard commented:
Filed a follow-up issue for defining and implementing metrics pings for these APIs

https://github.com/mozilla/fxa/issues/883

lmorchard commented:
Alright, I think I waved the white flag on resolving a few things within this already giant PR. Rebased against master and squashed. 

I'm seeing one last test failure related to QR codes on the content server. Hopefully that goes away?

lmorchard commented:
Welp. I'm going to stop touching this for awhile, get some dinner, and assume that the QR code failure is transient. 🤷‍♂️

lmorchard commented:
Re-ran tests after waiting most of a day and they passed. Guessing I ran us into some usage limit on the QR code stuff.

lmorchard commented:
Giving this one more tweak, because I think I worked out a fix for #887

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,201 @@
+  async function handleAuth(auth) {
+    const scope = ScopeSet.fromArray(auth.credentials.scope);
+    if (! scope.contains(SUBSCRIPTIONS_MANAGEMENT_SCOPE)) {
+      throw error.invalidToken('Invalid authentication scope in token');
```
@rfk can probably chip in here on the semantics of the errors, I suspect a better error is [invalidScopes](https://github.com/lmorchard/fxa/blob/96a000ea6e702e7474e7b016e692b19cc7db9d0d/packages/fxa-auth-server/lib/error.js#L1010).

lmorchard replied:
Just went ahead and changed this to invalidScopes.

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,201 @@
+      throw error.invalidToken('Invalid authentication scope in token');
+    }
+    const uid = auth.credentials.user;
+    const account = await db.account(uid);
```
Is `email` already on `auth.credentials`?

lmorchard replied:
IIRC, it was for a sessionToken but not an oauthToken

shane-tomlinson replied:
It looks like `email` should be set for an OAuth request in hapi-fxa-oauth here:

https://github.com/mozilla/hapi-fxa-oauth/blob/master/index.js#L107

The info set there comes from the OAuth server's /verify endpoint here:

https://github.com/mozilla/fxa-auth-server/blob/master/fxa-oauth-server/lib/routes/verify.js#L26

So it might be there and if so, we'd avoid an extra DB lookup here.

lmorchard replied:
Hmm, I hadn't been able to put those pieces together before, I think. Went ahead and simplified this

https://github.com/mozilla/fxa/pull/819/files#diff-eedc9e44d94c523680530e9c145c9496R26

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,201 @@
+
+  async function handleAuth(auth) {
+    const scope = ScopeSet.fromArray(auth.credentials.scope);
+    if (! scope.contains(SUBSCRIPTIONS_MANAGEMENT_SCOPE)) {
```
@rfk can clarify here too, we may want to differentiate between read and write access. URL scopes grant [read-write access](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/docs/scopes.md#url-scopes) by default, though some of these endpoints seem like they only require read-access. If we decide on differentiating between the two, that can be in a follow-on PR.

lmorchard replied:
Yeah, I think this is what I filed #882 for. If we're not using `profile:email` as the scope for reading subscription status, then we'll need a different one. We definitely don't want any client other than the management pages to have access to write APIs for subscription stuff.

rfk replied:
This seems reasonable to me to start with; [this doc](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/fxa-oauth-server/docs/scopes.md) has some suggested notation for distinguishing read-vs-write scopes, bit I tink having completely separate scopes for "view the user's subscription as part of their profile data" and "manage the user's subscription state" makes sense.  IIUC you would *never* expect an RP to be calling the `/subscription` APIs directly, even in read-only mode, right?

lmorchard replied:
> IIUC you would never expect an RP to be calling the /subscription APIs directly, even in read-only mode, right?

Yes, the only subscription-related APIs used by RPs live on the profile server, to fetch the user's set of relevant subscription capabilities. These subscription management APIs are only for use by our own subscription management pages.

> When requesting a grant for SUBSCRIPTIONS_MANAGEMENT_SCOPE we should ensure the client_id making the request is actually allowed to request the grant.

This should be a part of #881

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,201 @@
+        },
+        response: {
+          schema: isA.object().keys({
+            plan_id: validators.subscriptionsPlanId.required(),
```
Thanks for adding the validators.

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,201 @@
+      },
+      handler: async function (request) {
+        log.begin('subscriptions.listPlans', request);
+        await handleAuth(request.auth);
```
I wonder if there is a way to make `handleAuth` an [auth strategy](https://github.com/lmorchard/fxa/blob/96a000ea6e702e7474e7b016e692b19cc7db9d0d/packages/fxa-auth-server/lib/server.js#L341) so it doesn't need to be called in each of these routes. I'm not sure whether the required scopes can be passed as data to the auth-strategy somehow.

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/index.js:
```diff
@@ -86,7 +89,8 @@ module.exports = function (
     util,
-    recoveryKey
+    recoveryKey,
+    subscriptions
```
As mentioned above, can we add a config feature flag that says whether this feature is enabled and only add these routes if enabled? Note, this might be a good use of @philbooth's feature flagging, we could flip the feature on dynamically.

lmorchard replied:
Probably better to gate with a configuration flag on deployment, since setting up products & clients needs config tweaks too. I don't think we'd get much from dynamic flagging.

lmorchard replied:
https://github.com/mozilla/fxa/pull/819/commits/0f09f92c619439ffe6d522cc4c8a5c972b6bf70b

philbooth replied:
Also, ref: #475, feature-flagging is not set up in the auth server yet.

lmorchard commented on packages/fxa-auth-server/config/index.js:
```diff
@@ -569,6 +569,66 @@ const conf = convict({
+    },
+  },
+  subscriptions: {
+    managementClientId: {
```
https://github.com/mozilla/fxa/pull/819/commits/0f09f92c619439ffe6d522cc4c8a5c972b6bf70b

lmorchard commented:
Implemented a `config.subscriptions.enabled` flag and rebased against master. I have follow up issues open for issues around tokens & scopes (#881, #882), so it might be good to continue iteration there.

lmorchard commented:
Okay, another tweak to simplify `handleAuth` in the subscription routes. Also rebased against m again rather than keep them separate

lmorchard commented:
Hmm, seeing some test failures somehow related to sending email - maybe one of the intermittent issues we noticed elsewhere?

rfk commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -76,6 +76,12 @@ see [`mozilla/fxa-js-client`](https://github.com/mozilla/fxa-js-client).
+    * [GET /oauth/subscriptions/active (:lock: oauthToken)](#get-subscriptionsactive)
+    * [POST /oauth/subscriptions/active (:lock: oauthToken)](#post-subscriptionsactive)
+    * [DELETE /oauth/subscriptions/active/{subscriptionId} (:lock: oauthToken)](#delete-subscriptionsactivesubscriptionid)
+    * [POST /oauth/subscriptions/updatePayment (:lock: oauthToken)](#post-subscriptionsupdatepayment)
```
API design nit: it's not obvious to me why these live under `/oauth/subscriptions` rather than, say, `/account/subscriptions` or even `/subscriptions`.  My perspective is that the routes under `/oauth` are specifically for *managing oauth state* such as tokens and codes, which I'm not sure this is.  But anyway, no big deal if this structure makes sense to you, just something I noticed.

shane-tomlinson replied:
I suggested this, I think for API clarity it makes sense to me for everything authorized by an OAuth token to be under the /oauth path.

lmorchard replied:
Yeah, this was a change per review feedback from last week: https://github.com/mozilla/fxa/pull/819#pullrequestreview-227656302

rfk commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -3008,6 +3020,32 @@ Returns SMS status for the current user.
+#### POST /oauth/subscriptions/active
+
+:lock: authenticated with OAuth bearer token
+Subscribe the user to a plan using a payment token.
```
docs nit: "payment token" is not defined anywhere in this document, is there some other document that it could link out to with more details?

lmorchard replied:
It's an opaque token representing a payment method and consent to pay. Defined by [subhub](https://github.com/mozilla/subhub), which in turn takes it from [Stripe.js](https://stripe.com/docs/stripe-js/reference#stripe-create-token). But, subhub is a private project and we're trying to stay more agnostic to the underlying payment vendor. TL;DR: No great docs to link to about this

rfk commented on packages/fxa-auth-server/lib/routes/account.js:
```diff
@@ -1225,10 +1236,24 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push,
+                  for (const subscription of subscriptions) {
+                    const { subscriptionId } = subscription;
+                    await subhub.cancelSubscription(uid, subscriptionId);
+                    await db.deleteAccountSubscription(uid, subscriptionId);
```
What's the worst thing that could happen if we crash after `cancelSubscription`, or after `deleteAccountSubscription`?  Do we risk leaving the account in a state where it cannot be successfully deleted if the user retries again later?

rfk commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  const SUBSCRIPTIONS_MANAGEMENT_SCOPE =
+    'https://identity.mozilla.com/apps/subscriptions';
```
design nit: since this is an "internal" scope used by FxA itself, I wonder if it should be something like "/acct/subscriptions" rather than "/apps/subscriptions" for a bit of conceptual separation.  I don't think we really have any prior art for this though.

shane-tomlinson replied:
> I wonder if it should be something like "/acct/subscriptions" rather than "/apps/subscriptions" for a bit of conceptual separation. I don't think we really have any prior art for this though.

I left some thoughts [over here](https://github.com/mozilla/fxa/issues/882#issuecomment-485750529).

rfk commented on packages/fxa-auth-server/lib/subhub.js:
```diff
@@ -0,0 +1,315 @@
+    }), {});
+  }
+
+  const SubHubAPI = createBackendServiceAPI(log, config, 'subhub', {
```
As an aside, I'm happy to see this being useful for additional backend service calls :-)

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,204 @@
+        const plans = await subhub.listPlans();
+        const selectedPlan = plans.filter(p => p.plan_id === planId)[0];
+        if (! selectedPlan) {
+          throw error.unknownSubscriptionPlan(planId);
```
If this happens, what happens to the paymentToken, IIUC, the payment is already established and the user will be getting charged (possibly repeatedly?). Do we need to notify subhub to cancel the payment?

lmorchard replied:
As I understand it, the payment token represents a method of payment and consent to pay - i.e. not yet a completed payment.

It doesn't actually result in payment until we've successfully passed it to subhub and then subhub passes it back to the payment processor.

shane-tomlinson replied:
> It doesn't actually result in payment until we've successfully passed it to subhub and then subhub passes it back to the payment processor.

Good, let's verify just to make sure. @bbangert can you confirm ^^^, or is there someone from IT we can verify this with?

lmorchard replied:
I mean, I can verify it insofar as this is how we've been discussing it up til now and it's now the subhub API is spec'd to work.

shane-tomlinson commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,204 @@
+    },
+    {
+      method: 'DELETE',
+      path: '/oauth/subscriptions/active/{subscriptionId}',
```
Will all subscription cancellations occur on pages FxA controls or can the RP cancel a subscription?

lmorchard replied:
All subscription cancellations occur on FxA pages. RPs never have access to manage subscriptions.

lmorchard commented on packages/fxa-auth-server/docs/api.md:
```diff
@@ -3008,6 +3020,32 @@ Returns SMS status for the current user.
+:lock: authenticated with OAuth bearer token
+Returns a list of available subscription plans.
+
+#### GET /oauth/subscriptions/active
```
No, there will not be an `/inactive` resource. This is just what I named the user's active subscriptions as distinct from available plans

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,204 @@
+        const { planId, paymentToken } = request.payload;
+
+        // Find the selected plan and get its product ID
+        const plans = await subhub.listPlans();
```
These APIs are never called by RPs - they're only ever called by our own subscription management page on its own domain.

> how do we prevent RPs from subscribing users to a plan

RPs don't subscribe users to plans - only our subscription management pages can do that. And that only after the user consents to pay and produces a payment token.

> if this endpoint is called with a paymentToken, the payment is already established with the payment provider and we are too late.

The token can only be obtained through a JS widget we're hosting on the separate-domain payment pages. RPs cannot do this.

lmorchard commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -0,0 +1,204 @@
+      options: {
+        auth: {
+          payload: false,
+          strategy: 'oauthToken'
```
It looks like [the profile server](https://github.com/mozilla/fxa/blob/master/packages/fxa-profile-server/lib/server/web.js#L94) uses a different oauth scheme implementation than [the auth server](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/lib/server.js#L335). Also, I can't really tell where / how the profile implementation enforces scope and I'm leery of trying to change what the auth server does at a global level.

> Really, I'd just like to get rid of handleAuth.

Can we possibly spin this into a follow up issue?

lmorchard commented:
Went ahead and tweaked the scopes for read/write auth APIs and read-only profile APIs per #882. Rebased against master.

lmorchard commented on packages/fxa-auth-server/lib/customs.js:
```diff
@@ -107,11 +107,15 @@ module.exports = function (log, error) {
-    }
-    if (clonePayload.oldAuthPW) {
-      delete clonePayload.oldAuthPW;
+    const fieldsToOmit = [
```
Things that shouldn't be here from this PR? Or is this good to spin off as a follow-up?

lmorchard commented:
Merging while all the lights are green
